include/slb/detail/invoke.hpp

```diff
@@ -0,0 +1,389 @@
+                              std::declval<Tn>()...)))> {
+  using type = decltype(
+      (std::declval<T1>().get().*std::declval<F&>())(std::declval<Tn>()...));
+};
+
+template <typename F, typename T1, typename... Tn>
+struct mem_fun_ptr_result<ptr_tag,
+                          F(T1, Tn...),
+                          decltype((
+                              void)(((*std::declval<T1>()).*std::declval<F&>())(
+                              std::declval<Tn>()...)))> {
+  using type = decltype(
+      ((*std::declval<T1>()).*std::declval<F&>())(std::declval<Tn>()...));
+};
+
+////////////////////////////////////////////////////////////////////////////////
+template <typename T>
+class mem_fun_ptr;
+
+template <typename T, typename C>
+class mem_fun_ptr<T C::*> {
+  T C::*pm;
+
+private:
+  // `(t1.*pm)(t2, ..., tN)` when `pm` is a pointer to data member of a class
+  // `C` and `is_base_of_v<T, remove_reference_t<decltype(t1)>>` is `true`;
+  template <typename R, typename T1, typename... Tn>
+  static constexpr R call(ref_tag, T C::*pm, T1&& t1, Tn&&... tn) noexcept(
+      noexcept((std::forward<T1>(t1).*pm)(std::forward<Tn>(tn)...))) {
+    return (std::forward<T1>(t1).*pm)(std::forward<Tn>(tn)...);
+  }
+
+  // `(t1.get().*pm)(t2, ..., tN)` when `pm` is a pointer to data member of a
+  // class `C` and `remove_cvref_t<decltype(t1)>` is a specialization of
+  // `reference_wrapper`;
+  template <typename R, typename T1, typename... Tn>
+  static constexpr R
+  call(ref_wrapper_tag,
+       T C::*pm,
+       T1&& t1,
+       Tn&&... tn) noexcept(noexcept((t1.get().*pm)(std::forward<Tn>(tn)...))) {
+    return (t1.get().*pm)(std::forward<Tn>(tn)...);
+  }
+
+  // `((*t1).*pm)(t2, ..., tN)` when `pm` is a pointer to data member of a
+  // class `C` and `t1` does not satisfy the previous two items;
+  template <typename R, typename T1, typename... Tn>
+  static constexpr R call(ptr_tag, T C::*pm, T1&& t1, Tn&&... tn) noexcept(
+      noexcept(((*std::forward<T1>(t1)).*pm)(std::forward<Tn>(tn)...))) {
+    return ((*std::forward<T1>(t1)).*pm)(std::forward<Tn>(tn)...);
+  }
+
+public:
+  constexpr mem_fun_ptr(T C::*pm) noexcept : pm(pm) {}
+
+  template <typename T1,
+            typename... Tn,
+            typename PT1 = typename p0704<T, T1>::type,
+            typename Tag = typename dispatch_mem_ptr<C, PT1>::type,
+            typename R =
+                typename mem_fun_ptr_result<Tag, T C::*(PT1&&, Tn&&...)>::type,
+            typename = typename apply_clang_workaround<C, T, T1>::type>
+  constexpr R operator()(T1&& t1, Tn&&... tn) const noexcept(noexcept(
+      call<R>(Tag{}, pm, std::forward<PT1>(t1), std::forward<Tn>(tn)...))) {
+    return call<R>(Tag{}, pm, std::forward<PT1>(t1), std::forward<Tn>(tn)...);
+  }
+};
+
+} // namespace invoke_impl
+
+// We don't use `std::is_function<T>` here since some libstdc++ versions fail to
+// classify function types with noexcept specifiers.
+
+template <typename T>
+struct is_function {
```

Comment:
With https://github.com/mpark/slb/pull/24 merged, we should use `detail::lib::is_function` instead, right?